statsmodels/distributions/joint_distribution.py

```diff
@@ -0,0 +1,236 @@
+import numpy as np
+from scipy import stats
+from abc import ABC, abstractmethod
+
+
+class Copula(ABC):
+    r"""A generic Copula class meant for subclassing.
+
+    Parameters
+    ----------
+    seed : {None, int, `numpy.random.Generator`}, optional
+        If `seed` is None the `numpy.random.Generator` singleton is used.
+        If `seed` is an int, a new ``Generator`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``Generator`` instance then that instance is
+        used.
+
+    Notes
+    -----
+    A function :math:`\phi` on :math:`[0, \infty]` is the Laplace-Stieltjes
+    transform of a distribution function if and only if :math:`\phi` is
+    completely monotone and :math:`\phi(0) = 1` [2]_.
+
+    The following algorithm for sampling a ``d``-dimensional exchangeable
+    Archimedean copula with generator :math:`\phi` is due to Marshall, Olkin
+    (1988) [1]_, where :math:`LS^{−1}(\phi)` denotes the inverse
+    Laplace-Stieltjes transform of :math:`\phi`.
+
+    From a mixture representation with respect to :math:`F`, the following
+    algorithm may be derived for sampling Archimedean copulas, see [1]_.
+
+    1. Sample :math:`V \sim F = LS^{−1}(\phi)`.
+    2. Sample i.i.d. :math:`X_i \sim U[0,1], i \in \{1,...,d\}`.
+    3. Return:math:`(U_1,..., U_d)`, where :math:`U_i = \phi(−\log(X_i)/V), i
+       \in \{1, ...,d\}`.
+
+    Detailed properties of each copula can be found in [3]_.
+
+    Instances of the class can access the attributes: ``rng`` for the random
+    number generator (used for the ``seed``).
+
+    **Subclassing**
+
+    When subclassing `Copula` to create a new copula, ``__init__`` and
+    ``random`` must be redefined.
+
+    * ``__init__(seed=None)``: If the sampler
+      does not take advantage of a ``seed``, this parameter can be omitted.
+    * ``random(n)``: draw ``n`` from the copula.
+
+    References
+    ----------
+    .. [1] Marshall AW, Olkin I. “Families of Multivariate Distributions”,
+      Journal of the American Statistical Association, 83, 834–841, 1988.
+    .. [2] Marius Hofert. "Sampling Archimedean copulas", Universität Ulm, 2008.
+    .. [3] Harry Joe. "Dependence Modeling with Copulas", Monographs on
+      Statistics and Applied Probability 134, 2015.
+
+    """
+
+    @abstractmethod
+    def __init__(self, seed=None):
+        self.rng = np.random.default_rng(seed)
+
+    @abstractmethod
+    def random(self, n):
```

Comment:
Needs to accept an argument `random_state` that can handle a `random_state` or `BitGenerator` object.

Reply:
This is like what scipy does.  

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.rvs.html

Seed shouldn't be part of the class initialization.

Reply:
ok sure. I was not sure about the architecture here. If you think something else would be better in general let me know.

Reply:
For the accepted values, I propose to not talk about `RandomState ` as we do in `stats.qmc` to strive people to not use it. Let me know if I should do otherwise.

Reply:
It seems fine to me to prefer `Generator` objects.  The problem is that a `RandomState` has to be default until statsmodels requires NumPy 1.17.

Reply:
Is there somewhere a function to check the random state? Something like `scipy._lib._util.check_random_state`. I could not find anything like this.